4828 Toggle Fritekstvedlegg i fritekstbrev

fredrik-oh opened the pull request:
https://jira.adeo.no/browse/MELOSYS-4828

Dependent PRs:
https://github.com/navikt/melosys-web/pull/1581
https://github.com/navikt/melosys-dokgen/pull/191

Commit:
WIP, lagt til fritekstvedleggSkjema.tsx. Gjenstår visning av fritekstvedlegg i vedleggTable, forhåndsvisning av fritekstvedlegget samt integrasjon mot melosys-api for sending av brev med fritekstvedlegg

Commit:
La til støtte for å legge til, slette og endre fritekstvedlegg i fritekstbrev

TODO: Generell feilhåndtering for fritekstvedlegg, samt forhåndsvisning av vedlegget

Commit:
La til bekreftelsesvindu for sletting av fritekstvedlegg, la til snapshot tester for enkelte komponenter i vedleggvelger, oppdaterte .eslintrc

Skulle gjerne hatt bedre snapshot dekning i vedleggvelger mappen, men per nå har jeg ingen løsning på genererte feltverdier som settes på child components (id, htmlfor). Tenker å høre i frontend slacken om noen har funnet en god måte å løse dette, men utifra hva jeg kan se er det ingen enkel støtte for dette i jest

Commit:
Ryddet opp i import rekkefølge

IsaSalemHame commented on src/felleskomponenter/sideDialog/sendBrev/brevVedlegg.tsx:
```diff
@@ -5,21 +5,44 @@ import { FysiskDokument } from "Domene";
   setValgteVedlegg: (valgteVedlegg: FysiskDokument[]) => void;
+  fritekstvedlegg: Fritekstvedlegg[];
+  redigerFritekstvedlegg: (index: number) => void;
+  slettFritekstvedlegg: (index: number) => void;
```
Note4myself. Hva er index her?

fredrik-oh replied:
array index på fritekstvedlegg som ligger i sendbrev state

IsaSalemHame replied:
ja, skjønte det etterpå. glemte å fjerne kommentar. index høres så vagt ut hvis man ikke har en klar oversikt over hvordan ting henger sammen. Men samtidig har jeg ikke så mye bedre forslag :( Kunne kanskje ha funnet brev basert på tittel, men det er fort gjort å ha to like titler..

IsaSalemHame commented on src/felleskomponenter/sideDialog/sendBrev/fritekstvedleggSkjema.tsx:
```diff
@@ -0,0 +1,58 @@
+          </Nav.Column>
+        </Nav.Row>
+        <Skjema.HTMLEditor feltNavn={`felt.${felt.kode}_FRITEKST.feltVerdi`} />
+        <Nav.Row>
```
Skulle man hatt en "Forhåndsvis"-knapp også? Da slipper man å lagre, for så å forhåndsvise. Vet ikke hvor stort behov det vil være for det, men...
I hovedbrevet kan man alltids forhåndsvise, selv om man ikke er ferdig med å skrive.. 

En annen løsning kan være å ikke slette raden fra vedlegglista når man redigerer? Idk. Igjen, jeg vet ikke om hvor stort behov det er.

![image](https://user-images.githubusercontent.com/7471428/191250807-c1899f74-52f3-4c57-a906-29e136ac436b.png)

fredrik-oh replied:
Tror jeg går for å ikke fjerne vedlegget fra listen når vi redigerer

IsaSalemHame commented on src/felleskomponenter/vedleggvelger/vedleggTable.tsx:
```diff
@@ -1,87 +1,50 @@
+            redigerFritekstvedlegg={redigerFritekstvedlegg}
+            slettFritekstvedlegg={slettFritekstvedlegg}
+            key={vedlegg.tittel}
+            index={index}
```
Vi har ikke et bedre navn enn "index"?

IsaSalemHame commented on src/felleskomponenter/vedleggvelger/fritekstvedleggRow.tsx:
```diff
@@ -0,0 +1,62 @@
+        <Mui.Knapp
+          type="flat"
+          ikon={Ikoner.BlyantBlack}
+          onClick={() => redigerFritekstvedlegg && redigerFritekstvedlegg(index)}
```
Bug: Hvis du har to vedlegg, prøver å redigere den ene, for så å trykke rediger på neste også, så slettes ene vedlegget.

fredrik-oh replied:
Fikser dette ved å ikke fjerne vedlegget fra listen når vi redigerer

IsaSalemHame commented on src/felleskomponenter/sideDialog/sendBrev/sendBrev.tsx:
```diff
@@ -286,13 +357,31 @@ const SendBrev = ({
         />
       )}
 
+      {fritekstvedleggFelt &&
```
Noe føles rart her... 
Ene er at det er to like knapper, men i to forskjellige komponenter: `Legg til vedlegg` og `Legg til fritekstvedlegg`. 
Slik det er nå, er i tillegg FritekstvedleggSkjema avhengig av BrevVedlegg på grunn av listen over fritekstvedlegg som har blitt lagt til. I en rar verden med en brevtype i backend der man har fritekstvedlegg, men ikke vedlegg, vil man altså ikke kunne se vedlegget man har lagt på. 
Foreslår at tabellen over vedlegg som har blitt valgt/lagt til kommer seg ut på en eller annen måte så Fritekstvedlegg ikke er avhengig av BrevVedlegg. Eventuelt kan du flette fritekstvedleggFelt inn i BrevVedlegg.

IsaSalemHame commented on src/felleskomponenter/vedleggvelger/__snapshots__/slettFritekstvedleggModal.test.tsx.snap:
```diff
@@ -0,0 +1,73 @@
+    </section>
+    <button
+      aria-label="Lukk"
+      className="lukknapp lukknapp--overstHjorne modal__lukkknapp--shake"
```
kkk

IsaSalemHame commented on src/felleskomponenter/vedleggvelger/slettFritekstvedleggModal.tsx:
```diff
@@ -0,0 +1,47 @@
+      isOpen
+      shouldCloseOnOverlayClick
+      onRequestClose={onRequestClose}
+      // @ts-ignore
```
Jeg fikk ikke noe fel uten `@ts-ignore` :O

IsaSalemHame commented on src/felleskomponenter/sideDialog/sendBrev/fritekstvedleggSkjema.tsx:
```diff
@@ -0,0 +1,58 @@
+            Lagre fritekstvedlegg
+          </Nav.Hovedknapp>
+          <Nav.Knapp mini onClick={resetFritekstvedlegg}>
+            Avbryt
```
Bug: Hvis du lagrer fritekstvedlegg, redigerer og trykker Avbryt, så slettes hele fritekstvedlegget.

IsaSalemHame commented on src/felleskomponenter/sideDialog/sendBrev/sendBrev.tsx:
```diff
@@ -171,9 +183,29 @@ const SendBrev = ({
+    try {
+      return await dokumenterOperations.forhandsvisBrevV2(behandlingID, data);
+    } catch (e) {
+      // TODO: Feilhåndtering
```
TODO

fredrik-oh replied:
Skulle vært notification fra github når du prøver å opprette en PR med todo kommentarer i endringssettet 😅

IsaSalemHame replied:
tror du kan få intellij til å varsle deg

IsaSalemHame commented on src/felleskomponenter/sideDialog/sendBrev/brevVedlegg.tsx:
```diff
@@ -5,21 +5,44 @@ import { FysiskDokument } from "Domene";
+  fritekstvedlegg: Fritekstvedlegg[];
+  redigerFritekstvedlegg: (index: number) => void;
+  slettFritekstvedlegg: (index: number) => void;
+  lagPdfUrl: (index: number) => Promise<string | false>;
```
Ser ut som det bare blir brukt ifm. fritekstvedleggvisning? I så fall kall det noe annet. Kanskje lagFritekstPdfUrl, som du allerede har kalt det et annet sted.

Commit:
Refaktoring av VedleggVelger, splittet opp i to felleskomponenter VedleggVelger og VedleggTable

Commit:
Fikset flere feil ved redigering av eksisterende fritekstvedlegg, la til varsel for feilet forhåndsvisning av fritekstvedlegg

Commit:
Ryddet opp i lagFritekstPdfUrl request body

IsaSalemHame commented on src/felleskomponenter/sideDialog/sendBrev/sendBrev.tsx:
```diff
@@ -279,20 +351,40 @@ const SendBrev = ({
+        <VedleggTable
           valgteVedlegg={valgteVedlegg}
+          fritekstvedlegg={fritekstvedlegg}
+          redigerFritekstvedlegg={redigerFritekstvedlegg}
```
Det er sikkert ikke nødvendig for denne omgang, men hadde vært fint å få det på plass en eller annen gang. Når man redigerer et fritekstvedlegg, for så å trykke på Rediger på en annen så burde bruker få varsel om at fritekstvedlegget som allerede var i redigeringsmodus ikke var blitt lagret og spørre om hen faktisk vil fortsette.

IsaSalemHame commented on src/felleskomponenter/vedleggTable/vedleggTable.tsx:
```diff
@@ -0,0 +1,64 @@
+                <FritekstvedleggRow
+                  fritekstvedlegg={vedlegg}
+                  redigerFritekstvedlegg={redigerFritekstvedlegg}
+                  slettFritekstvedlegg={slettFritekstvedlegg}
```
Ved sletting av fritekstvedlegg bør også redigeringsboksen bli lukket, dersom den er åpen for samme vedlegg som man sletter. Nå fører det til en liten bug hvis man gjør det på riktig index.

IsaSalemHame replied:
(gammel kommentar fra i går som kom inn med reviewen)

IsaSalemHame commented on src/felleskomponenter/vedleggTable/vedleggTable.tsx:
```diff
@@ -0,0 +1,64 @@
+                  fritekstvedlegg={vedlegg}
+                  redigerFritekstvedlegg={redigerFritekstvedlegg}
+                  slettFritekstvedlegg={slettFritekstvedlegg}
+                  key={vedlegg.tittel}
```
Key må være unik, tittel er ikke det. key=index?

fredrik-oh replied:
index gir warning og eslint feil mener jeg :(

barstadEkstern commented on src/felleskomponenter/vedleggTable/vedleggTable.tsx:
```diff
@@ -0,0 +1,64 @@
+  return (
+    <>
+      <Nav.Typo.Element>{label}</Nav.Typo.Element>
+      {(valgteVedlegg.length > 0 || (fritekstvedlegg && fritekstvedlegg.length > 0)) && (
```
Burde vi sjekke valgteVedlegg også før vi sjekker lengden?

fredrik-oh replied:
Sjekker om fritekstvedlegg er definert siden det er en optional param

barstadEkstern reviewed (approved):
Ser ikke noe spesielt utover det som er nemnt. La inn en liten kommentar som kanskje kan fikses om du synest det gir mening.

Commit:
Fikset et par bugs i sendBrev.slettFritekstvedlegg

Commit:
La til scrolling og max height på text editor, justerte forhåndsvisning av fritekstvedlegg.

Dersom saksbehandler redigerer et fritekstvedlegg vises den redigerte tittel og/eller tekst i stedet for de "lagrede" verdiene.

Commit:
Merge branch 'master' into MELOSYS-4828_Fritekstvedlegg_til_brev

# Conflicts:
#	package-lock.json
#	package.json

IsaSalemHame reviewed (approved):
Ser bra ut! Fiks testene som feiler, så er det good.

Commit:
Fjernet test som krever redux-form i kontekst

Commit:
La til manglende snapshot

IsaSalemHame reviewed (approved):
:100:

Commit:
La til toggle på fritekstvedlegg
